stlab/concurrency/default_executor.hpp

```diff
@@ -247,56 +240,84 @@ inline auto queue_size() {
     // The test cannot run with less than two cores
     return std::max(2u, std::thread::hardware_concurrency());
 #else
-    return std::thread::hardware_concurrency();
+    return std::max(1u, std::thread::hardware_concurrency());
 #endif
 }
 
 class notification_queue {
     using lock_t = std::unique_lock<std::mutex>;
-    std::deque<task<void()>> _q;
-    std::mutex _queue_mutex;
-    std::condition_variable* _ready{nullptr};
 
-public:
-    void set_context(std::condition_variable& cv) {
-        _ready = &cv;
+    struct element_t {
+        unsigned _priority;
+        task<void()> _task;
+
+        template <class F>
+        element_t(F&& f, unsigned priority) : _priority{priority}, _task{std::forward<F>(f)} { }
+
+        struct greater {
+            bool operator()(const element_t& a, const element_t& b) const {
+                return b._priority < a._priority;
+            }
+        };
+    };
+
+    std::vector<element_t> _q; // can't use priority queue because top() is const
```

Comment:
What is the reason not to use a deque here? Is the locality of a vector better for a heap structure than a deque?

Reply:
For a priority queue (heap) a vector is more efficient. With a heap you always add/remove elements from the back, so we have less heap allocations than a dequeue (with the tradeoff that it stays at the high-water mark, fixable if it mattered and you get faster, less complex indexing.